pkg/dataobj/consumer/partition_processor.go

```diff
@@ -338,6 +302,52 @@ func (p *partitionProcessor) processRecord(record *kgo.Record) {
 	p.lastModified = p.clock.Now()
 }
 
+// flushAndCommit flushes the builder and, if successful, commits the offset
+// of the last record processed. It expects that the last record processed
+// was also the last record appended to the builder. If not, data loss can
+// occur should the consumer restart or a partition rebalance occur
+func (p *partitionProcessor) flushAndCommit() error {
+	if err := p.flush(); err != nil {
+		return fmt.Errorf("failed to flush: %w", err)
+	}
+	if err := p.commit(); err != nil {
+		return fmt.Errorf("failed to commit offset: %w", err)
+	}
+	return nil
+}
+
+// flush builds a complete data object from the builder, uploads it, records
+// it in the metastore, and emits an object written event to the events topic.
+func (p *partitionProcessor) flush() error {
+	obj, closer, err := p.builder.Flush()
+	if err != nil {
+		level.Error(p.logger).Log("msg", "failed to flush builder", "err", err)
+		return err
+	}
+	defer closer.Close()
+
+	objectPath, err := p.uploader.Upload(p.ctx, obj)
+	if err != nil {
+		level.Error(p.logger).Log("msg", "failed to upload object", "err", err)
+		return err
+	}
+
+	minTime, maxTime := p.builder.TimeRange()
```

Comment:
the time range needs to be read before the builder flushes, otherwise the builder is empty and these times are both zero

Reply:
Good point! Could you attempt to add a test for this? If its too tricky, create a follow up ticket and I'll take care of it today

Reply:
Done! 😄